File: configurator/src/utils/sysex.ts
// Config-over-SysEx v1 codec. Mirror of libfp/src/sysex.rs — keep in sync.
//
// Envelope: F0 7D 46 50 01 <7-bit-packed payload> F7
// Packed payload (8-bit domain): u16 BE length prefix + postcard bytes.

export const SYSEX_START = 0xf0;
export const SYSEX_EOX = 0xf7;
export const SYSEX_HEADER = new Uint8Array([0x7d, 0x46, 0x50, 0x01]);

// Pack 8-bit bytes into 7-bit MIDI data bytes: per group of up to 7 input
// bytes, one MSB byte (bit i = top bit of byte i) followed by the low 7 bits
// of each byte.
export function pack7bit(src: Uint8Array): Uint8Array {
  const dst = new Uint8Array(src.length + Math.ceil(src.length / 7));
  let written = 0;
  for (let group = 0; group < src.length; group += 7) {
    const groupLen = Math.min(7, src.length - group);
    const msbIndex = written++;
    dst[msbIndex] = 0;
    for (let i = 0; i < groupLen; i++) {
      const byte = src[group + i];
      dst[msbIndex] |= (byte >> 7) << i;
      dst[written++] = byte & 0x7f;
    }
  }
  return dst;
}

export function unpack7bit(src: Uint8Array): Uint8Array {
  const dst = new Uint8Array(src.length - Math.ceil(src.length / 8));
  let written = 0;
  for (let group = 0; group < src.length; group += 8) {
    const groupLen = Math.min(8, src.length - group);
    if (groupLen < 2) {
      throw new Error("Truncated 7-bit packed data");
    }
    const msb = src[group];
    for (let i = 1; i < groupLen; i++) {
      const byte = src[group + i];
      if ((byte & 0x80) !== 0 || (msb & 0x80) !== 0) {
        throw new Error("Invalid byte in 7-bit packed data");
      }
      dst[written++] = byte | (((msb >> (i - 1)) & 1) << 7);
    }
  }
  return dst.slice(0, written);
}

// Wraps postcard bytes into a complete config SysEx frame.
export function buildConfigFrame(payload: Uint8Array): Uint8Array {
  const plain = new Uint8Array(payload.length + 2);
  plain[0] = (payload.length >> 8) & 0xff;
  plain[1] = payload.length & 0xff;
  plain.set(payload, 2);

  const packed = pack7bit(plain);
  const frame = new Uint8Array(1 + SYSEX_HEADER.length + packed.length + 1);
  frame[0] = SYSEX_START;
  frame.set(SYSEX_HEADER, 1);
  frame.set(packed, 1 + SYSEX_HEADER.length);
  frame[frame.length - 1] = SYSEX_EOX;
  return frame;
}

// Extracts the postcard bytes from a complete F0..F7 frame. Returns null for
// frames that are not ours (foreign SysEx) or fail validation.
export function parseConfigFrame(frame: Uint8Array): Uint8Array | null {
  if (
    frame.length < 2 + SYSEX_HEADER.length ||
    frame[0] !== SYSEX_START ||
    frame[frame.length - 1] !== SYSEX_EOX
  ) {
    return null;
  }
  for (let i = 0; i < SYSEX_HEADER.length; i++) {
    if (frame[1 + i] !== SYSEX_HEADER[i]) {
      return null;
    }
  }
  const packed = frame.slice(1 + SYSEX_HEADER.length, frame.length - 1);
  let plain: Uint8Array;
  try {
    plain = unpack7bit(packed);
  } catch {
    return null;
  }
  if (plain.length < 2) {
    return null;
  }
  const payloadLength = (plain[0] << 8) | plain[1];
  if (payloadLength !== plain.length - 2) {
    return null;
  }
  return plain.slice(2);
}

File: libfp/src/sysex.rs
//! Config-over-SysEx v1 wire format shared between firmware and configurator.
//!
//! Envelope: `F0 7D 46 50 01 <7-bit-packed payload> F7`
//!
//! The packed payload (in the 8-bit domain, before packing) is a u16 BE length
//! prefix followed by the postcard-serialized `ConfigMsgIn`/`ConfigMsgOut`.
//! The configurator mirrors this codec in `configurator/src/utils/sysex.ts` —
//! keep both sides in sync.

pub const SYSEX_START: u8 = 0xF0;
pub const SYSEX_EOX: u8 = 0xF7;

/// Bytes between `F0` and the packed payload: manufacturer ID 0x7D
/// (non-commercial), "FP" device signature, protocol version 1. A registered
/// manufacturer ID can replace 0x7D later by changing only this constant (and
/// its TS mirror).
pub const SYSEX_HEADER: [u8; 4] = [0x7D, 0x46, 0x50, 0x01];

/// Max postcard payload size (unchanged from the WebUSB protocol).
pub const MAX_PAYLOAD_SIZE: usize = 512;
/// Payload plus the u16 BE length prefix.
pub const MAX_PLAIN_SIZE: usize = MAX_PAYLOAD_SIZE + 2;
/// `packed_len(MAX_PLAIN_SIZE)`.
pub const MAX_PACKED_SIZE: usize = packed_len(MAX_PLAIN_SIZE);
/// Full frame: F0 + header + packed payload + F7.
pub const MAX_SYSEX_FRAME: usize = 1 + SYSEX_HEADER.len() + MAX_PACKED_SIZE + 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SysexError {
    /// A packed byte had its top bit set.
    InvalidByte,
    /// Destination buffer too small.
    BufferTooSmall,
    /// Packed input ended in the middle of a group in an invalid way.
    Truncated,
}

/// Packed size for `n` plain bytes: one MSB byte per group of up to 7.
pub const fn packed_len(n: usize) -> usize {
    n + n.div_ceil(7)
}

/// Pack 8-bit bytes into 7-bit MIDI data bytes. For each group of up to 7
/// input bytes, emits one MSB byte (bit i = top bit of byte i) followed by the
/// low 7 bits of each byte. Returns the number of bytes written.
pub fn pack_7bit(src: &[u8], dst: &mut [u8]) -> Result<usize, SysexError> {
    if dst.len() < packed_len(src.len()) {
        return Err(SysexError::BufferTooSmall);
    }
    let mut written = 0;
    for group in src.chunks(7) {
        let msb_idx = written;
        dst[msb_idx] = 0;
        written += 1;
        for (i, &byte) in group.iter().enumerate() {
            dst[msb_idx] |= (byte >> 7) << i;
            dst[written] = byte & 0x7F;
            written += 1;
        }
    }
    Ok(written)
}

/// Inverse of [`pack_7bit`]. Returns the number of plain bytes written.
pub fn unpack_7bit(src: &[u8], dst: &mut [u8]) -> Result<usize, SysexError> {
    let mut written = 0;
    for group in src.chunks(8) {
        // A group must contain the MSB byte plus at least one data byte.
        if group.len() < 2 {
            return Err(SysexError::Truncated);
        }
        let msb = group[0];
        for (i, &byte) in group[1..].iter().enumerate() {
            if byte & 0x80 != 0 || msb & 0x80 != 0 {
                return Err(SysexError::InvalidByte);
            }
            if written >= dst.len() {
                return Err(SysexError::BufferTooSmall);
            }
            dst[written] = byte | (((msb >> i) & 1) << 7);
            written += 1;
        }
    }
    Ok(written)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn roundtrip(plain: &[u8]) {
        let mut packed = [0u8; packed_len(MAX_PLAIN_SIZE)];
        let packed_size = pack_7bit(plain, &mut packed).unwrap();
        assert_eq!(packed_size, packed_len(plain.len()));
        assert!(packed[..packed_size].iter().all(|b| b & 0x80 == 0));
        let mut unpacked = [0u8; MAX_PLAIN_SIZE];
        let plain_size = unpack_7bit(&packed[..packed_size], &mut unpacked).unwrap();
        assert_eq!(&unpacked[..plain_size], plain);
    }

    #[test]
    fn empty_roundtrip() {
        roundtrip(&[]);
    }

    #[test]
    fn small_roundtrips() {
        roundtrip(&[0x00]);
        roundtrip(&[0xFF]);
        roundtrip(&[0x80, 0x7F, 0x00, 0xFF, 0x01, 0xFE, 0xAA]);
        roundtrip(&[0x80, 0x7F, 0x00, 0xFF, 0x01, 0xFE, 0xAA, 0x55]);
    }

    #[test]
    fn max_size_roundtrip() {
        let mut plain = [0u8; MAX_PLAIN_SIZE];
        for (i, byte) in plain.iter_mut().enumerate() {
            *byte = (i % 256) as u8;
        }
        roundtrip(&plain);
    }

    #[test]
    fn unpack_rejects_high_bit() {
        let mut dst = [0u8; 8];
        assert_eq!(
            unpack_7bit(&[0x00, 0x80], &mut dst),
            Err(SysexError::InvalidByte)
        );
    }

    #[test]
    fn unpack_rejects_lone_msb_byte() {
        let mut dst = [0u8; 8];
        assert_eq!(unpack_7bit(&[0x01], &mut dst), Err(SysexError::Truncated));
    }

    #[test]
    fn frame_size_consts() {
        assert_eq!(MAX_PACKED_SIZE, 588);
        assert_eq!(MAX_SYSEX_FRAME, 594);
    }
}
